File: app/lib/cycle_ig/smart_health_link.dart
import 'dart:convert';

import 'package:menstrudel/cycle_ig/shlink_crypto.dart';

class SmartHealthLink {
  final String url;
  final String key;
  final String flag;
  final String? label;
  final int? exp;
  final int v;

  const SmartHealthLink({
    required this.url,
    required this.key,
    this.flag = 'U',
    this.label,
    this.exp,
    this.v = 1,
  });

  Map<String, dynamic> toPayload() {
    return {
      'url': url,
      'key': key,
      'flag': flag,
      if (label != null) 'label': label,
      if (exp != null) 'exp': exp,
      'v': v,
    };
  }

  String toBareShlink() {
    final encoded = ShlinkCrypto.base64UrlNoPadding(
      utf8.encode(jsonEncode(toPayload())),
    );
    return 'shlink:/$encoded';
  }

  String toViewerLink({String viewerBase = 'https://cycle.fhir.me/view'}) {
    return '$viewerBase#${toBareShlink()}';
  }

  static SmartHealthLink parse(String input) {
    final marker = input.indexOf('shlink:/');
    if (marker < 0) {
      throw FormatException('No shlink:/ fragment found.');
    }
    final encoded = input.substring(marker + 'shlink:/'.length);
    final payload =
        jsonDecode(utf8.decode(ShlinkCrypto.decodeBase64UrlNoPadding(encoded)))
            as Map<String, dynamic>;

    return SmartHealthLink(
      url: payload['url'] as String,
      key: payload['key'] as String,
      flag: payload['flag'] as String? ?? '',
      label: payload['label'] as String?,
      exp: payload['exp'] as int?,
      v: payload['v'] as int? ?? 1,
    );
  }
}

File: app/src/main/java/com/kollectivemobile/euki/cycleig/SmartHealthLink.java
package com.kollectivemobile.euki.cycleig;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

import java.nio.charset.StandardCharsets;
import java.util.Base64;

public class SmartHealthLink {
    private static final Base64.Encoder B64_URL_ENCODER = Base64.getUrlEncoder().withoutPadding();
    private static final Base64.Decoder B64_URL_DECODER = Base64.getUrlDecoder();

    private SmartHealthLink() {
    }

    public static String composeViewerLink(String viewerBaseUrl, String fileUrl, String key, long exp, String label) {
        JsonObject payload = new JsonObject();
        payload.addProperty("url", fileUrl);
        payload.addProperty("key", key);
        payload.addProperty("flag", "U");
        payload.addProperty("label", label);
        payload.addProperty("exp", exp);
        payload.addProperty("v", 1);
        String shlink = "shlink:/" + b64Url(payload.toString().getBytes(StandardCharsets.UTF_8));
        return viewerBaseUrl.replaceAll("/+$", "") + "#" + shlink;
    }

    public static JsonObject parse(String input) {
        int index = input.indexOf("shlink:/");
        if (index < 0) {
            throw new IllegalArgumentException("no shlink:/ found");
        }
        String payload = input.substring(index + "shlink:/".length());
        return JsonParser.parseString(new String(b64UrlDecode(payload), StandardCharsets.UTF_8)).getAsJsonObject();
    }

    public static String b64Url(byte[] bytes) {
        return B64_URL_ENCODER.encodeToString(bytes);
    }

    public static byte[] b64UrlDecode(String value) {
        return B64_URL_DECODER.decode(value);
    }
}
